adding cascade learning blurb (#56)
jaybee84/ml-in-rd: content/03.prior-knowledge.cascadelearning.md

## Cascade Learning {.page_break_before}

Sometimes, due to paucity of data and its unstable predictive capabilities, one may feel the need to consult various machine learning methods and the predictions there-of to make an educated decision about the identified predictive patterns. 

> This assembly of multiple machine learning methods is called ensemble learning. In an ensemble learning workflow, one may use various parameters to maximize the learned information. 

Reviewer: Two sentences in one line

Reviewer: This doesn't appear to be resolved, but maybe i am looking at the wrong thing.

> Sometimes a consensus approach is ideal to select predictive features from a group of independently running models, other times one may coordinate various learning methods in such a way that one model uses information from another model in augmenting its predictive capability. 

Reviewer: I would change this to "A consensus approach is often...." to avoid this sentence starting with Sometimes so soon after another one

Random forest algorithms are a great example of ensemble learning where a consensus of the features selected through many different and independent decision trees are considered and ranked to make a final ranked list of important features that are predictive of a certain outcome (e.g. disease) [@doi:10.1186/1472-6947-13-134, @doi:10.1023/A:1010933404324].

> However, successful application of consensus based ensemble learning algorithms prefer what is known as "gold standard" data. These generally refer to well labeled cases where the diagnosis or label of a data point in the training dataset has very little uncertainty (or "noise") associated with it [@doi:10.1093/jamia/ocw028]. 

Reviewer: Two sentences in one line. 

Also, I am having a bit of trouble following the first sentence. Should be "prefer" be something more like: "requires"?

> In most cases of rare disease, due to the inherent nature of being less defined, the symptoms as well as any underlying biology comes with a reasonable amount of uncertainty (or "noise") leading to a "silver standard" dataset[@doi:10.1109/TNNLS.2013.2292894, @doi:10.1093/jamia/ocw028].

Reviewer: I have never heard of silver dataset, interesting! I did recently hear someone referring to bronze-standard methods :) 

My experience is that rare diseases lack a good gold standard not due to noise, but rather your initial point in this section (a paucity of data). I'm also not sure if "noise" is the right way to describe phenotypic heterogeneity/uncertainty. In the case of complex diseases, the phenotypes are genuinely variable, not meaningless (and I tend to think of noise as meaningless data - not true variance).

Author: I see what you mean. Actually I thought the same until I came across this paper(https://pubmed.ncbi.nlm.nih.gov/24808033/). It describes various kinds of noise in the context of machine learning like this:
"... anything that obscures the relationship between the features of an instance and its class. .... noise is also described as consisting of nonsystematic errors ......... In the literature, two types of noise are distinguished: feature(or attribute) and class noises. On the one hand, feature noise affects the observed values of the feature, e.g. by adding a small Gaussian noise to each feature during measurement. On the other hand, class noise alters the observed labels assigned to instances, e.g. by incorrectly setting a negative label on a positive instance in binary classification." 

In our case the term "label-noise" may be more appropriate in describing the true phenotypic uncertainty in the features that seem close to the "Noisy not at random" (NNAR) model described in the paper.

> This kind of data begets modifications in ensemble learning methods where one method leverages findings from another method to capture stable patterns existing in the "silver standard" data and reduce uncertainty.

Reviewer: I think this could be combined with the next sentence. "This challenge can be resolved with a type of ensemble learning called cascade learning which is ...."

Such methods fall into the category of cascade learning.

> An example of cascade learning that showed robustness in view of noisy data was implemented to identify rare disease patients from electronic health records from the general population [@pmid:30815073].

Reviewer: This is an interesting project, though that table (Table 1, where they pick some top-lists of patients to be evaluated) has self-assessment trap written all over it.

Author: :) yeah their patient assessment is not the best... but the case study still describes the principles of cascade learning which is why I decided to include this study here.. still looking to see if we can find other examples of work to cite but not coming across many cascade learning papers in rare disease :)

This implementation consisted of three steps each employing a different independent learning algorithm: (1) feature extraction to assign text words (from Pubmed literature) to diagnosis using word2vec [@arXiv:1301.3781v3], (2) preliminary prediction using an ensemble of decision trees with penalization for excessive tree-depth, (3) prediction refinement using similarity of data points to de-noise sample labels and reiterating step (2).
In this implementation the algorithm was able to identify rare data due to the robustness conferred by the independence of the feature extraction step and the prediction refinement step from the preliminary classification of the labeled dataset.

> The classification step capitalized upon the information learned by the label prediction step preceding it, and helped the prediction refinement step following it through its enhanced classification accuracy.

Reviewer: I don't know that this paper actually shows this in a systematic way, but maybe i just missed this when skimming.

Author: Can you elaborate a bit on this? 

This was my takeaway from the paper, but happy to reassess if there are specific metrics I missed

Reviewer: Maybe I missed it, but I didn't see any comparison of metrics  describing the predictive accuracy bwithout and with the prediction refinement step.

## Cascade Learning {.page_break_before}

> Sometimes, due to paucity of data and its unstable predictive capabilities, one may feel the need to consult various machine learning methods and the predictions there-of to make an educated decision about the identified predictive patterns, collectively called _ensemble learning_. 

Reviewer 2: I would probably rephrase `unstable predictive capabilities` – this strikes me as a bit of jargon.

> Sometimes, due to paucity of data and its unstable predictive capabilities, one may feel the need to consult various machine learning methods and the predictions there-of to make an educated decision about the identified predictive patterns, collectively called _ensemble learning_. 

Reviewer 2: ```suggestion
Sometimes, due to paucity of data and its unstable predictive capabilities, one may feel the need to consult various machine learning methods and their predictions to make an educated decision about the identified predictive patterns, collectively called _ensemble learning_. 
```
Would saying "to trust the results" get the same point across as "identified predictive patterns" ?

Author: yes.. that definitely sounds better and simplifies the sentence :)

> In an ensemble learning workflow, one may use various parameters to maximize the learned information. 

Reviewer 2: I'm not sure what this sentence does for the intended audience.

Author: This was to segway into the next sentence where we differentiate between ensembles-using-consensus vs ensembles-using-cascades

Reviewer 2: Gotcha – in that case, I think it needs more context to connect the two.

> A consensus approach is often ideal to select predictive features from a group of independently running models, other times one may coordinate various learning methods in such a way that one model uses information from another model in augmenting its predictive capability. 

Reviewer 2: ```suggestion
A consensus approach is often ideal to select predictive features from a group of independently running models. 
At other times, one may coordinate various learning methods in such a way that one model uses information from another model in augmenting its predictive capability. 
```

In most cases of rare disease, due to the inherent nature of being less defined, the symptoms as well as any underlying biology comes with a reasonable amount of uncertainty (or "label-noise") leading to a _silver standard_ dataset[@doi:10.1109/TNNLS.2013.2292894, @doi:10.1093/jamia/ocw028].
This kind of data begets modifications in ensemble learning methods where one method leverages findings from another method to capture stable patterns existing in the _silver standard_ data and reduce uncertainty.
Such methods fall into the category of cascade learning.
Among classifiers, cascade learning has been widely used in image recognition where initially a small subset of image features are used to classify images (e.g. features identifying a face like eyes, nose, mouth). 
The initial classification is then augmented by more complex features and algorithms like AdaBoost ( _boosting_ ) that weight the various features are implemented to detect the content of the image (e.g. features identifying a human face like relative distance between eyes etc.)[@doi:10.1109/CVPR.2001.990537, @doi:10.1007/978-3-540-75175-5_16, @doi:10.1109/ICPR.2004.1334680].

> In rare diseases, a variant of cascade learning that showed robustness in view of uncertainty in the data was implemented to identify rare disease patients from electronic health records from the general population [@pmid:30815073].

Reviewer 2: What do you mean by `in view of uncertainty` here?

Author: This was to refer to the data labels in rare disease that had uncertainty associated with them (label-noise). However, I added in the previous lines later on and now this sentence looks out of context... need to smoothen it out a bit :)

This implementation consisted of three steps each employing a different independent learning algorithm: (1) feature extraction to assign text words (from Pubmed literature) to diagnosis using word2vec [@arXiv:1301.3781v3], (2) preliminary prediction using an ensemble of decision trees with penalization for excessive tree-depth, (3) prediction refinement using similarity of data points to resolve sample labels and reiterating step (2).

> In this implementation the algorithm was able to identify rare data due to the robustness conferred by the independence of the feature extraction step and the prediction refinement step from the preliminary classification of the labeled dataset.

Reviewer 2: ```suggestion
In this implementation the algorithm was able to identify rare disease patients due to the robustness conferred by the independence of the feature extraction step and the prediction refinement step from the preliminary classification of the labeled dataset.
```
Does this retain your meaning?

## Cascade Learning {.page_break_before}

> Sometimes, due to paucity of data and its unstable predictive capabilities, one may feel the need to consult various machine learning methods and their predictions to make an educated decision about the identified predictive patterns, collectively called _ensemble learning_. 

Reviewer 3: I feel like this is too long of an introduction & it might be better to explain very briefly why one might want an ensemble (a single predictor may not adequately capture the data) + discuss approaches that use models with the same parameters w/ different starting points (i.e., RF, k-means clustering) or those that use multiple distinct algorithms with different assumptions/etc.

In an ensemble learning workflow, one may use various parameters to maximize the learned information. 
A consensus approach is often ideal to select predictive features from a group of independently running models. 
At other times, one may coordinate various learning methods in such a way that one model uses information from another model in augmenting its predictive capability. 
Random forest algorithms are a great example of ensemble learning where a consensus of the features selected through many different and independent decision trees are considered and ranked to make a final ranked list of important features that are predictive of a certain outcome (e.g. disease) [@doi:10.1186/1472-6947-13-134, @doi:10.1023/A:1010933404324].

> For the ensemble of decision trees, e.g. random forest, bootstrap aggregation (or _bagging_) helps in forming the consensus[@doi:10.1214/aos/1031689014, @doi:10.1177/2045894019890549, @doi:10.1016/S0031-3203(02)00169-3].

Reviewer 3: ```suggestion
For the ensemble of decision trees, e.g. random forest, bootstrap aggregation (or _bagging_) helps in forming the consensus[@doi:10.1214/aos/1031689014, @doi:10.1177/2045894019890549, @doi:10/btzfh6].
```
use shortdoi b/c parents in doi are special characters

However, successful application of consensus based ensemble learning algorithms requires what is known as "gold standard" data. 
These generally refer to well labeled cases where the diagnosis or label of a data point in the training dataset has very little uncertainty (or "label-noise") associated with it [@doi:10.1093/jamia/ocw028]. 
In most cases of rare disease, due to the inherent nature of being less defined, the symptoms as well as any underlying biology comes with a reasonable amount of uncertainty (or "label-noise") leading to a _silver standard_ dataset[@doi:10.1109/TNNLS.2013.2292894, @doi:10.1093/jamia/ocw028].
This kind of data begets modifications in ensemble learning methods where one method leverages findings from another method to capture stable patterns existing in the _silver standard_ data and reduce uncertainty.

> Such methods fall into the category of cascade learning.

Reviewer 3: Title is cascade learning, but cascade learning doesn't come up until here. Perhaps better to use ensemble learning as the title? Or separate into its own section.

Author: :) Yes.. this started off as a section for cascade learning, but based on @allaway and @jaclyn-taroni 's reviews and the clear need for introducing ensemble learning before delving into cascade learning, I have been adding more ensemble learning introduction to this PR. But missed adding this information to the PR description. Apologies for the confusion!
